homeassistant/components/tplink/entity.py

```diff
@@ -0,0 +1,70 @@
+"""Common code for tplink."""
+from __future__ import annotations
+
+from typing import Any, Callable, cast
+
+from kasa import SmartDevice
+
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .coordinator import TPLinkDataUpdateCoordinator
+
+
+def async_refresh_after(func: Callable) -> Callable:
+    """Define a wrapper to refresh after."""
+
+    async def _async_wrap(
+        self: CoordinatedTPLinkEntity, *args: Any, **kwargs: Any
+    ) -> None:
+        await func(self, *args, **kwargs)
+        await self.coordinator.async_request_refresh_without_children()
+
+    return _async_wrap
+
+
+class CoordinatedTPLinkEntity(CoordinatorEntity):
+    """Common base class for all coordinated tplink entities."""
+
+    coordinator: TPLinkDataUpdateCoordinator
+
+    def __init__(
+        self, device: SmartDevice, coordinator: TPLinkDataUpdateCoordinator
+    ) -> None:
+        """Initialize the switch."""
+        super().__init__(coordinator)
+        self.device: SmartDevice = device
+
+    @property
+    def data(self) -> dict[str, Any]:
+        """Return data from DataUpdateCoordinator."""
+        data: dict[str, Any] = self.coordinator.data
+        return data
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return cast(str, self.device.device_id)
```

Comment:
Is it possible to have 2 entities for a single device, ie 2 sensors? In that case this wouldn't work right? 

Also should we just assign this to `self._attr_unique_id` in the constructor ?

Reply:
The property gets overwritten in `sensor.py`

Reply:
Switched to `_attr_unique_id`

[`648004b` (#56632)](https://github.com/home-assistant/core/pull/56632/commits/648004bbde6706669b43746d1e33390433caaf0b)